### ISSUE TITLE
`clippy:: map_entry` incorrectly warned on code with a `contains_key` and no `insert`

### ISSUE THREAD

**Issue by erickt** [rust-lang/rust-clippy#14449]
### Summary

Fuchsia just did a dry run roll of Rust https://github.com/rust-lang/rust/commit/5d85a714b10a9121d2ec5115f39e904174f804e1, and it's incorrectly flagging the `clippy::map_entry` lint in at least two bits of our code that use `contains_key`, but do not have a corresponding `insert`:

* [src/connectivity/network/netstack3/src/bindings/filter/conversion.rs](https://cs.opensource.google/fuchsia/fuchsia/+/main:src/connectivity/network/netstack3/src/bindings/filter/conversion.rs;l=392)

```
error: usage of `contains_key` followed by `insert` on a `HashMap`
   --> ../../src/connectivity/network/netstack3/src/bindings/filter/conversion.rs:392:17
    |
392 | /                 if self.routine_types.contains_key(&name) {
393 | |                     self.expect_routine_resolved(&name, routine_type, rule_id)
394 | |                 } else {
395 | |                     self.convert_routine(
...   |
401 | |                 }
    | |_________________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
    = note: `-D clippy::map-entry` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::map_entry)]`
```

* [src/devices/lib/driver-static-checks/src/checks.rs](https://cs.opensource.google/fuchsia/fuchsia/+/main:src/devices/lib/driver-static-checks/src/checks.rs;l=150):

```
error: usage of `contains_key` followed by `insert` on a `HashMap`
   --> ../../src/devices/lib/driver-static-checks/src/checks.rs:150:9
    |
150 | /         if json_category.contains_key(&device_category.subcategory) {
151 | |             Ok(StaticCheckPass {})
152 | |         } else {
153 | |             Err(StaticCheckFail  {
...   |
158 | |             })
159 | |         }
    | |_________^
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_entry
    = note: `-D clippy::map-entry` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::map_entry)]`
```

### Lint Name

clippy:: map_entry

### Reproducer

So far I haven't been able to reproduce it outside our code, I'll update if I figure out one.

### Version

```text
This was built in our CI/CQ, so it's a little hard to get `rustc -Vv`, but it's with the nightly rustc version https://github.com/rust-lang/rust/commit/5d85a714b10a9121d2ec5115f39e904174f804e1.
```

**Comment by elenaf9** [comment]
We are running into the same issue:

- [libp2p/rust-libp2p/protocols/gossipsub/src/peer_score.rs](https://github.com/libp2p/rust-libp2p/blob/f4322c5a31b06dd2f678376e66fcd9dd675f80ea/protocols/gossipsub/src/peer_score.rs#L113-L117): 
  (_note that it's_ `params.topics` _and not_ `self.topics` _in the first statement)_
  ```sh
  error: usage of `contains_key` followed by `insert` on a `HashMap`
     --> protocols/gossipsub/src/peer_score.rs:113:9
      |
  113 | /         if params.topics.contains_key(&topic_hash) {
  114 | |             Some(self.topics.entry(topic_hash).or_default())
  115 | |         } else {
  116 | |             self.topics.get_mut(&topic_hash)
  117 | |         }
      | |_________^
      |
  ```
 
- [libp2p/rust-libp2p/protocols/gossipsub/src/behaviour.rs](https://github.com/libp2p/rust-libp2p/blob/da003b031c2f5d024d51d832ab0c2575356885fd/protocols/gossipsub/src/behaviour.rs#L1816-L1830):
  ```sh
  error: usage of `contains_key` followed by `insert` on a `HashMap`
      --> protocols/gossipsub/src/behaviour.rs:1816:9
       |
  1816 | /         if self.mesh.contains_key(&message.topic) {
  1817 | |             tracing::debug!("Sending received message to user");
  1818 | |             self.events
  1819 | |                 .push_back(ToSwarm::GenerateEvent(Event::Message {
  ...    |
  1829 | |             return;
  1830 | |         }
       | |_________^
  ```

## Version

clippy 0.1.87 (45165c82a4 2025-04-01) (current beta)

**Comment by meithecatte** [comment]
Minimal example ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=7b6d0b42855af6e272cebe8e999d3ebd)):

```rust
use std::collections::BTreeMap;
pub struct Meow {
    map: BTreeMap<String, String>,
}

impl Meow {
    fn pet(&self, _key: &str, _v: u32) -> u32 {
        42
    }
}

pub fn f(meow: &Meow, x: String) {
    if meow.map.contains_key(&x) {
        let _ = meow.pet(&x, 1);
    } else {
        let _ = meow.pet(&x, 0);
    }
}
```

**Comment by meithecatte** [comment]
git bisect identifies 18616dcd4fce257e2a994914d8680ea4ed0098fe as the first bad commit. cc @profetia
